File: Distribution-Grid-EV-CA/R_Yanning/02_04_get_EV trips.R
library(data.table)

load.files <- function(directory) {
  files.all <- list.files(path=directory)
  output <- data.table()
  for(f in files.all) {
    hold <- fread(paste0(directory,'/',f))
    output <- rbind(output,hold)
  }
  return(output)
}

TAZ.dist.LD <- load.files(directory='data/mobility data/TAZ distance/parsed_100000_LDPTM')
TAZ.dist.ETM <- load.files(directory='data/mobility data/TAZ distance/parsed_100000_ETM')

TAZ.dist.LD.short <- TAZ.dist.LD[,c("from","to","distance")]
TAZ.dist.LD.short <- unique(TAZ.dist.LD.short)
TAZ.dist.ETM.short <- TAZ.dist.ETM[,c("from","to","distance")]
TAZ.dist.ETM.short <- unique(TAZ.dist.ETM.short)

## don't touch! ####
load.trips <- function() {
  files.all <- list.files()
  output <- data.table()
  for(f in files.all) {
    hold <- fread(f)
    output <- rbind(output,hold)
  }
  return(output)
}

setwd('C:/Nina/Research (large files)/CSTDM/SDPTM')
all.trips.sd <- load.trips()
all.trips.sd.short <- all.trips.sd[,c("SerialNo","Person","Tour","Trip","DPurp","I","J","Mode","Dist","Time")]
setwd('C:/Users/nina3/Box Sync/Distribution Grid/Distribution Grid EV CA')
saveRDS(all.trips.sd.short,file='data/mobility data/CSTDM processed/SDPTM trips short.rds')

#### start from here if already saved the short version
setwd('C:/Users/nina3/Box Sync/Distribution Grid/Distribution Grid EV CA')


# short distance EV trips
EVhh.sd <- fread('data/mobility data/CSTDM processed/EVhh_new_SDPTM_sample42.csv')
all.trips.sd.short <- readRDS('data/mobility data/CSTDM processed/SDPTM trips short.rds')
EV.trips.sd <- all.trips.sd.short[SerialNo%in%EVhh.sd$hhID & (Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3'),
                            c("SerialNo","Person","Tour","Trip","DPurp","I","J","Mode","Dist","Time")]
EV.trips.sd <- merge(x=EV.trips.sd,y=EVhh.sd[,c("hhID","year")],
                     by.x="SerialNo",by.y="hhID")
#EV.trips.sd <- readRDS('data/mobility data/CSTDM processed/EV trips_new_SDPTM_sample42.rds')

aggregate.purpose <- data.table('ChargeType'=c('H','W','W','P','P','P','P','P','P','P','P'),
                                    'DPurp'=c('O','W','P','S','H','T','C','L','R','K','Z'))
EV.trips.sd <- merge(x=EV.trips.sd,y=aggregate.purpose,by='DPurp')

saveRDS(EV.trips.sd,file='data/mobility data/CSTDM processed/EV trips_new_SDPTM_sample42.rds')

# long distance EV trips
EVhh.ld <- fread('data/mobility data/CSTDM processed/EVhh_new_LDPTM_sample42.csv')
all.trips.ld <- fread('C:/Nina/Research (large files)/CSTDM/LDPTM/LDPTM_Trips.csv')
EV.trips.ld <- all.trips.ld[SerialNo%in%EVhh.ld$hhID & (Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3'),
                            c("SerialNo","Person","Tour","Trip","DPurp","I","J","Mode","Dist","Time")]
EV.trips.ld <- merge(x=EV.trips.ld,y=EVhh.ld[,c("hhID","year")],
                     by.x="SerialNo",by.y="hhID")

aggregate.purpose <- data.table('ChargeType'=c('W','H','P','W','P'),
                                'DPurp'=c('Bus','VFR','Rec','Com','Oth'))
EV.trips.ld <- merge(x=EV.trips.ld,y=aggregate.purpose,by='DPurp')

saveRDS(EV.trips.ld,file='data/mobility data/CSTDM processed/EV trips_new_LDPTM_sample42.rds')

### just to check if the distances from the matrix all match
EV.trips.ld <- readRDS(file='data/mobility data/CSTDM processed/EV trips_new_LDPTM_sample42.rds')
EV.trips.ld.dist <- merge(x=EV.trips.ld,y=TAZ.dist.LD.short,
                          by.x=c("I","J"),by.y=c("from","to"),
                          all.x=T,allow.cartesian=T)
summary(EV.trips.ld.dist$distance)

trips.ld.dist <- merge(x=all.trips.ld,y=TAZ.dist.LD.short,
                          by.x=c("I","J"),by.y=c("from","to"),
                          all.x=T,allow.cartesian=T)
summary(trips.ld.dist$distance)

trips.ld.dist[,distance_mile:=distance*0.000621371192]
head(trips.ld.dist)
trips.ld.dist[,delta_dist:=Dist-distance_mile]
summary(trips.ld.dist$delta_dist) # difference range from -153.65 to 34.57 miles, mean -21.43

# long distance AccEgr EV trips - No Dist
EVhh.ld.accegr <- fread('data/mobility data/CSTDM processed/EVhh_new_LDPTM_AccEgr_sample42.csv')
all.trips.ld.accegr <- fread('C:/Nina/Research (large files)/CSTDM/LDPTM/LDPTM_AccEgr.csv')
unique(all.trips.ld.accegr$Mode)
unique(all.trips.ld.accegr$AccMode)
unique(all.trips.ld.accegr$DPurp)

EV.trips.ld.accegr <- all.trips.ld.accegr[SerialNo%in%EVhh.ld.accegr$hhID 
                                          & ((DPurp=='Access'&AccMode=='Drive_Park')|(OPurp=='Egress'&EgrMode=='Rent_Car')),
                                          c("SerialNo","DPurp","I","J","Mode","AccMode","EgrMode","Time")]
EV.trips.ld.accegr <- merge(x=EV.trips.ld.accegr,y=EVhh.ld.accegr[,c("hhID","year")],
                            by.x="SerialNo",by.y="hhID")

unique(EV.trips.ld.accegr$DPurp)
aggregate.purpose <- data.table('ChargeType'=c('W','H','P','W','P','P'),
                                'DPurp'=c('Bus','VFR','Rec','Com','Oth','Access'))
EV.trips.ld.accegr <- merge(x=EV.trips.ld.accegr,y=aggregate.purpose,by='DPurp')

saveRDS(EV.trips.ld.accegr,file='data/mobility data/CSTDM processed/EV trips_new_LDPTM_AccEgr_sample42.rds')

EV.trips.ld.accegr.dist <- merge(x=EV.trips.ld.accegr,y=TAZ.dist.short,
                            by.x=c("I","J"),by.y=c("from","to"))
# only 5 trips can find a match in the distance matrix, might need to discard this category of trips


# EXT EV trips - No Dist
all.trips.ext <- fread('C:/Nina/Research (large files)/CSTDM/ETM/trips_Ext.csv')
EVhh.ext <- fread('data/mobility data/CSTDM processed/EVhh_new_ETM_sample42.csv')

EV.trips.ext <- all.trips.ext[SerialNo%in%EVhh.ext$hhID 
                              & (ActorType=='CarLong'|ActorType=='CarLocal')
                              & (Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3') 
                              & DPurp=='I',
                              c("SerialNo","Person","Tour","Trip","DPurp","I","J","Mode","ActorType","Time")]
EV.trips.ext <- merge(x=EV.trips.ext,y=EVhh.ext[,c("hhID","year")],
                      by.x="SerialNo",by.y="hhID")
EV.trips.ext[,ChargeType:="P"]
EV.trips.ext.dist <- merge(x=EV.trips.ext,y=TAZ.dist.ETM.short,
                           by.x=c("I","J"),by.y=c("from","to"))
EV.trips.ext.dist[,Dist:=distance/1609.344]
summary(EV.trips.ext.dist)

saveRDS(EV.trips.ext.dist,file='data/mobility data/CSTDM processed/EV trips_new_EXT_sample42.rds')


# y <- 2044
# EV.trips <- all.trips.sd[SerialNo%in%EVhh.sd[year<=y]$hhID & (Mode=='SOV'|Mode=='HOV2'|Mode=='HOV3'),
#                          c("SerialNo","Person","Tour","Trip","DPurp","J","Mode","Dist")]
# saveRDS(EV.trips,file=paste0('data/mobility data/CSTDM processed/EV trips_SDPTM_',y,'_sample42'))
# EV.trips <- readRDS(paste0('data/mobility data/CSTDM processed/EV trips_SDPTM_',y,'_sample42'))

File: Distribution-Grid-EV-CA/02_04_get_EV_trips.py
"""
02_04_get_EV_trips.py — Python port of ``02_04_get_EV trips.R``.

Filters the raw CSTDM trip tables down to EV-household trips for SDPTM, LDPTM,
LDPTM AccEgr, and ETM, then joins aggregate charging-purpose categories.
Outputs pickle files (``.pkl``) in place of the R ``.rds`` files.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

from common import fread_all_in_dir, read_rds_like, save_rds_like

SDPTM_DIR = "data/mobility_data/CSTDM/SDPTM"
LDPTM_TRIPS = "data/mobility_data/CSTDM/LDPTM/LDPTM_Trips.csv"
LDPTM_ACCEGR = "data/mobility_data/CSTDM/LDPTM/LDPTM_AccEgr.csv"
ETM_TRIPS = "data/mobility_data/CSTDM/ETM/trips_Ext.csv"
PARSED_LDPTM_DIR = "data/mobility_data/TAZ_distance/parsed_100000_LDPTM"
PARSED_ETM_DIR = "data/mobility_data/TAZ_distance/parsed_100000_ETM"

OUT_DIR = "data/mobility_data/CSTDM_processed"


# charging-purpose aggregation tables (from the R script)
AGG_SD = pd.DataFrame(
    {
        "ChargeType": ["H", "W", "W", "P", "P", "P", "P", "P", "P", "P", "P"],
        "DPurp":      ["O", "W", "P", "S", "H", "T", "C", "L", "R", "K", "Z"],
    }
)
AGG_LD = pd.DataFrame(
    {"ChargeType": ["W", "H", "P", "W", "P"], "DPurp": ["Bus", "VFR", "Rec", "Com", "Oth"]}
)
AGG_LD_AE = pd.DataFrame(
    {
        "ChargeType": ["W", "H", "P", "W", "P", "P"],
        "DPurp": ["Bus", "VFR", "Rec", "Com", "Oth", "Access"],
    }
)


def _load_taz_dist_short(directory: str) -> pd.DataFrame:
    df = fread_all_in_dir(directory)
    if df.empty:
        return df
    return df[["from", "to", "distance"]].drop_duplicates()


def main() -> None:
    os.makedirs(OUT_DIR, exist_ok=True)

    # TAZ distance matrices (short columns only)
    taz_dist_ld = _load_taz_dist_short(PARSED_LDPTM_DIR)
    taz_dist_etm = _load_taz_dist_short(PARSED_ETM_DIR)

    # --- SDPTM ---------------------------------------------------------------
    all_sd = fread_all_in_dir(SDPTM_DIR)
    cols_short = ["SerialNo", "Person", "Tour", "Trip", "DPurp", "I", "J", "Mode", "Dist", "Time"]
    all_sd_short = all_sd[cols_short].copy()
    save_rds_like(all_sd_short, os.path.join(OUT_DIR, "SDPTM trips short.pkl"))

    evhh_sd = pd.read_csv(os.path.join(OUT_DIR, "EVhh_new_SDPTM_sample42.csv"))
    ev_sd = all_sd_short[
        all_sd_short["SerialNo"].isin(evhh_sd["hhID"])
        & all_sd_short["Mode"].isin(["SOV", "HOV2", "HOV3"])
    ]
    ev_sd = pd.merge(
        ev_sd,
        evhh_sd[["hhID", "year"]],
        left_on="SerialNo",
        right_on="hhID",
    ).drop(columns=["hhID"])
    ev_sd = pd.merge(ev_sd, AGG_SD, on="DPurp")
    save_rds_like(ev_sd, os.path.join(OUT_DIR, "EV trips_new_SDPTM_sample42.pkl"))

    # --- LDPTM ---------------------------------------------------------------
    evhh_ld = pd.read_csv(os.path.join(OUT_DIR, "EVhh_new_LDPTM_sample42.csv"))
    all_ld = pd.read_csv(LDPTM_TRIPS)
    ev_ld = all_ld[
        all_ld["SerialNo"].isin(evhh_ld["hhID"])
        & all_ld["Mode"].isin(["SOV", "HOV2", "HOV3"])
    ][cols_short].copy()
    ev_ld = pd.merge(
        ev_ld, evhh_ld[["hhID", "year"]], left_on="SerialNo", right_on="hhID"
    ).drop(columns=["hhID"])
    ev_ld = pd.merge(ev_ld, AGG_LD, on="DPurp")
    save_rds_like(ev_ld, os.path.join(OUT_DIR, "EV trips_new_LDPTM_sample42.pkl"))

    if not taz_dist_ld.empty:
        ev_ld_dist = pd.merge(
            ev_ld,
            taz_dist_ld,
            left_on=["I", "J"],
            right_on=["from", "to"],
            how="left",
        )
        print(ev_ld_dist["distance"].describe())

    # --- LDPTM AccEgr --------------------------------------------------------
    evhh_ld_ae = pd.read_csv(os.path.join(OUT_DIR, "EVhh_new_LDPTM_AccEgr_sample42.csv"))
    all_ld_ae = pd.read_csv(LDPTM_ACCEGR)
    mask_ae = all_ld_ae["SerialNo"].isin(evhh_ld_ae["hhID"]) & (
        ((all_ld_ae["DPurp"] == "Access") & (all_ld_ae["AccMode"] == "Drive_Park"))
        | ((all_ld_ae["OPurp"] == "Egress") & (all_ld_ae["EgrMode"] == "Rent_Car"))
    )
    ev_ld_ae = all_ld_ae.loc[
        mask_ae,
        ["SerialNo", "DPurp", "I", "J", "Mode", "AccMode", "EgrMode", "Time"],
    ].copy()
    ev_ld_ae = pd.merge(
        ev_ld_ae, evhh_ld_ae[["hhID", "year"]], left_on="SerialNo", right_on="hhID"
    ).drop(columns=["hhID"])
    ev_ld_ae = pd.merge(ev_ld_ae, AGG_LD_AE, on="DPurp")
    save_rds_like(ev_ld_ae, os.path.join(OUT_DIR, "EV trips_new_LDPTM_AccEgr_sample42.pkl"))

    # --- ETM -----------------------------------------------------------------
    evhh_etm = pd.read_csv(os.path.join(OUT_DIR, "EVhh_new_ETM_sample42.csv"))
    all_etm = pd.read_csv(ETM_TRIPS)
    mask_etm = (
        all_etm["SerialNo"].isin(evhh_etm["hhID"])
        & all_etm["ActorType"].isin(["CarLong", "CarLocal"])
        & all_etm["Mode"].isin(["SOV", "HOV2", "HOV3"])
        & (all_etm["DPurp"] == "I")
    )
    ev_etm = all_etm.loc[
        mask_etm,
        ["SerialNo", "Person", "Tour", "Trip", "DPurp", "I", "J", "Mode", "ActorType", "Time"],
    ].copy()
    ev_etm = pd.merge(
        ev_etm, evhh_etm[["hhID", "year"]], left_on="SerialNo", right_on="hhID"
    ).drop(columns=["hhID"])
    ev_etm["ChargeType"] = "P"
    ev_etm_dist = pd.merge(
        ev_etm, taz_dist_etm, left_on=["I", "J"], right_on=["from", "to"]
    )
    ev_etm_dist["Dist"] = ev_etm_dist["distance"] / 1609.344
    save_rds_like(ev_etm_dist, os.path.join(OUT_DIR, "EV trips_new_EXT_sample42.pkl"))


if __name__ == "__main__":
    main()
